File: packages/dfos-protocol/src/chain/multikey.ts
/*

  MULTIKEY

  W3C Multikey encoding for Ed25519 public keys

*/

import { base58btc } from 'multiformats/bases/base58';

// Ed25519 public key multicodec: 0xed, varint-encoded as [0xed, 0x01]
const ED25519_PUB_PREFIX = new Uint8Array([0xed, 0x01]);

// Ed25519 private key multicodec: 0x1300, varint-encoded as [0x80, 0x26]
const ED25519_PRIV_PREFIX = new Uint8Array([0x80, 0x26]);

/** Ed25519 public key multicodec value */
export const ED25519_PUB_MULTICODEC = 0xed;

/** Ed25519 private key multicodec value */
export const ED25519_PRIV_MULTICODEC = 0x1300;

/**
 * Encode an Ed25519 public key as a W3C Multikey multibase string
 *
 * Format: 'z' + base58btc([0xed, 0x01] + publicKeyBytes)
 * Result starts with "z6Mk..."
 */
export const encodeEd25519Multikey = (publicKeyBytes: Uint8Array): string => {
  if (publicKeyBytes.length !== 32) {
    throw new Error(`expected 32-byte Ed25519 public key, got ${publicKeyBytes.length}`);
  }
  const prefixed = new Uint8Array(ED25519_PUB_PREFIX.length + publicKeyBytes.length);
  prefixed.set(ED25519_PUB_PREFIX);
  prefixed.set(publicKeyBytes, ED25519_PUB_PREFIX.length);
  return base58btc.encode(prefixed);
};

/**
 * Decode a Multikey multibase string to raw key bytes and codec
 *
 * Supports Ed25519 public keys (z6Mk... prefix)
 */
export const decodeMultikey = (multibase: string): { keyBytes: Uint8Array; codec: number } => {
  const bytes = base58btc.decode(multibase);

  if (bytes.length < 2) {
    throw new Error('multikey too short');
  }

  // Ed25519 public key
  if (bytes[0] === ED25519_PUB_PREFIX[0] && bytes[1] === ED25519_PUB_PREFIX[1]) {
    const keyBytes = bytes.slice(2);
    if (keyBytes.length !== 32) {
      throw new Error(`expected 32-byte Ed25519 public key, got ${keyBytes.length}`);
    }
    return { keyBytes, codec: ED25519_PUB_MULTICODEC };
  }

  // Ed25519 private key
  if (bytes[0] === ED25519_PRIV_PREFIX[0] && bytes[1] === ED25519_PRIV_PREFIX[1]) {
    const keyBytes = bytes.slice(2);
    if (keyBytes.length !== 32) {
      throw new Error(`expected 32-byte Ed25519 private key, got ${keyBytes.length}`);
    }
    return { keyBytes, codec: ED25519_PRIV_MULTICODEC };
  }

  throw new Error(
    `unsupported multikey codec: [0x${bytes[0]?.toString(16)}, 0x${bytes[1]?.toString(16)}]`,
  );
};

File: packages/dfos-cli/internal/protocol/multikey.go
package protocol

import (
	"fmt"

	"github.com/mr-tron/base58"
)

// EncodeMultikey encodes an Ed25519 public key as a Multikey multibase string.
// Format: "z" + base58btc(0xed 0x01 || pubkey)
func EncodeMultikey(pub []byte) string {
	raw := append([]byte{0xed, 0x01}, pub...)
	return "z" + base58.Encode(raw)
}

// DecodeMultikey decodes a Multikey multibase string to raw Ed25519 public key bytes.
func DecodeMultikey(multibase string) ([]byte, error) {
	if len(multibase) == 0 || multibase[0] != 'z' {
		return nil, fmt.Errorf("expected base58btc prefix 'z'")
	}
	raw, err := base58.Decode(multibase[1:])
	if err != nil {
		return nil, fmt.Errorf("base58 decode: %w", err)
	}
	if len(raw) < 2 || raw[0] != 0xed || raw[1] != 0x01 {
		return nil, fmt.Errorf("expected ed25519-pub multicodec prefix (0xed01)")
	}
	return raw[2:], nil
}
